components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -153,6 +282,114 @@ auto is_leaf_node(
     }
     return true;
 }
+
+auto get_sub_schema_tree_bitmap(
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs,
+        SchemaTree const& schema_tree
+) -> OUTCOME_V2_NAMESPACE::std_result<vector<bool>> {
+    auto sub_schema_tree_bitmap{vector<bool>(schema_tree.get_size(), false)};
+    for (auto const& [node_id, val] : node_id_value_pairs) {
+        if (node_id >= sub_schema_tree_bitmap.size()) {
+            return std::errc::result_out_of_range;
+        }
+        sub_schema_tree_bitmap[node_id] = true;
+
+        // Iteratively mark the parents as true
+        auto parent_id{schema_tree.get_node(node_id).get_parent_id()};
+        while (true) {
+            if (sub_schema_tree_bitmap[parent_id]) {
+                // Parent already set by other child
+                break;
+            }
+            sub_schema_tree_bitmap[parent_id] = true;
+            if (SchemaTree::cRootId == parent_id) {
+                break;
+            }
+            parent_id = schema_tree.get_node(parent_id).get_parent_id();
+        }
+    }
+    return sub_schema_tree_bitmap;
+}
+
+auto get_children_in_the_sub_schema_tree(
+        SchemaTreeNode const& node,
+        vector<bool> const& sub_schema_tree_bitmap
+) -> vector<SchemaTreeNode::id_t> {
+    vector<SchemaTreeNode::id_t> children;
+    for (auto const child_id : node.get_children_ids()) {
+        if (sub_schema_tree_bitmap[child_id]) {
+            children.push_back(child_id);
+        }
+    }
+    return children;
+}
+
+auto insert_kv_pair_json_map(
+        SchemaTreeNode const& node,
+        std::optional<Value> const& val,
+        nlohmann::json::object_t& json_map
+) -> bool {
+    auto const key_name{node.get_key_name()};
+    auto const type{node.get_type()};
+    if (false == val.has_value()) {
+        json_map.emplace(string{key_name}, nlohmann::json::object());
```

Comment:
How about extracting `string{key_name}` into a variable to avoid repeating it throughout this function?